merkle-tree/hash-set/src/lib.rs

```diff
@@ -0,0 +1,872 @@
+    ///
+    /// * `bytes` is casted into a reference of `HashSet` and used as
+    ///   storage for the struct.
+    /// * `capacity_indices` indicates the size of the indices table. It should
+    ///   already include a desired load factor and be greater than the expected
+    ///   number of elements to avoid filling the set too early and avoid
+    ///   creating clusters.
+    /// * `capacity_values` indicates the size of the values array. It should be
+    ///   equal to the number of expected elements, without load factor.
+    /// * `sequence_threshold` indicates a difference of sequence numbers which
+    ///   make elements of the has set expired. Expiration means that they can
+    ///   be replaced during insertion of new elements with sequence numbers
+    ///   higher by at least a threshold.
+    ///
+    /// # Purpose
+    ///
+    /// This method is meant to be used mostly in Solana programs to initialize
+    /// a new account which is supposed to store the hash set.
+    ///
+    /// # Safety
+    ///
+    /// This is highly unsafe. Ensuring the alignment and that the slice has
+    /// a correct size, which is able to fit the hash set, is the caller's
+    /// responsibility.
+    ///
+    /// Calling it in async context (or anywhere where the underlying data can
+    /// be moved in memory) is certainly going to cause undefined behavior.
+    pub unsafe fn from_bytes_init(
+        bytes: &mut [u8],
+        capacity_indices: I,
+        capacity_values: I,
+        sequence_threshold: usize,
+    ) -> &mut Self {
+        let hash_set_ref = &mut *(bytes.as_mut_ptr() as *mut Self);
+
+        hash_set_ref.capacity_indices = capacity_indices;
+        hash_set_ref.capacity_values = capacity_values;
+        hash_set_ref.sequence_threshold = sequence_threshold;
+
+        let offset = mem::size_of::<Self>();
+        hash_set_ref.indices =
+            NonNull::new(bytes.as_mut_ptr().add(offset) as *mut Option<I>).unwrap();
+
+        let offset = offset
+            + (mem::size_of::<Option<I>>()
+                * usize::try_from(hash_set_ref.capacity_indices).unwrap());
+        hash_set_ref.values =
+            NonNull::new(bytes.as_mut_ptr().add(offset) as *mut Option<HashSetCell>).unwrap();
+
+        hash_set_ref
+    }
+
+    // Insert a value into the hash set
+    pub fn insert(&mut self, value: &BigUint, sequence_number: usize) -> Result<(), HashSetError> {
+        for i in
+            0..usize::try_from(self.capacity_values).map_err(|_| HashSetError::IntegerOverflow)?
+        {
+            let i = I::try_from(i).map_err(|_| HashSetError::IntegerOverflow)?;
+            let probe_index = (value.clone() + i.to_biguint().unwrap() * i.to_biguint().unwrap())
+                % self.capacity_values.to_biguint().unwrap();
+            let probe_index = probe_index.to_usize().unwrap();
+            let index_bucket = unsafe { &mut *self.indices.as_ptr().add(probe_index) };
+
+            match index_bucket {
```

Comment:
I had a hard time understanding this match statement, because the nesting and code prevent me from seeing all the possible paths this statement can take at a glance. Github ui definitely didn't help either.

```
match index_bucket {
  /// Case is the normal operation of the hash set.
  /// If sequence number is expired overwrite old data
  /// If existing data value == insertion value throw error, value already exists.
  /// If sequence number is not expired continue with quadratic probing.
  Some(value_index) => {
     ....
     let index_bucket ...;
     if_sequence_number_is_expired_use_cell_and_overwrite_old_data(index_bucket)
  },
  /// Case can only be encountered when filling the hash set the first time, once all values have been used once this case ceases to occur.
  None => store_value_in_fresh_cell(),
 }
 ```